### DISCUSSION TITLE
Observing the height of a torso-attached point or frame?

### DISCUSSION THREAD

**Discussion by mariogpascual** [upkie/upkie#117] [Software] [answered]
Hi! Is it possible to extend the observation with some notion for Upkie's height, such as the coordinates of the center point of the upper torso cover? I've been studying the observers code but I don't see a way to access this information.

**Comment by stephane-caron** [comment, marked as answer]
I see two ways to interpret your question (not sure which one you mean). Both are doable but not currently implemented :wink: 

### 1) Getting ground truth info from the simulator

This would be a feature request for [Vulp](https://github.com/tasts-robots/vulp/) since that's where the simulator interface resides. I can give a bit more context on how to get there, feel free to open an issue in the [Vulp tracker](https://github.com/tasts-robots/vulp/issues).

### 2) Computing the coordinates of a given point

(Or more generally the pose of any robot frame with respect to the inertial frame.) This can be done here with a C++ or Python observer. A Python observer would be simpler to get started, *e.g.* following roughly these steps:

- Load a Pinocchio model, as done in the Pink balancer.
- Compute forward kinematics from sensory data, *i.e.* joint angles and IMU data.
- Assume the ground is flat and horizontal with respect to the inertial frame.
- Deduce from this data and assumption the locations of wheel contact points.
- Return height as the difference in z-coordinate (in the world frame) between the torso point and wheel contact points.

Contrary to getting ground truth from the simulator, having an observer would work both in sim and on real robots.

**Comment by mariogpascual** [comment]
Looks good, thank you :+1: I'm going for the second option. I opened issue tasts-robots/vulp#41 for the first one.
